### ISSUE TITLE
Stuck txpool txs not getting properly dropped/demoted

### ISSUE THREAD

**Issue by dankostiuk** [0xPolygon/polygon-edge#711]
### Description

It appears the txpool dropping or demoting logic isn't working as expected.

Today we had an incident on our mainnet where we saw many stuck queued and pending txs (see txpool_inspect response below) across all of our nodes. All stuck txs belonged to a single 'from' account - owned by us - and we had to resort to restarting all nodes in order to get our service up and running normally again.

Looking into our logs, it appears that the tx with nonce 707 attempted to be promoted, followed by a series of:
```
[ERROR] polygon: failed to apply tx: err=\"incorrect nonce\""
```

Eventually we see the messages indicating the dropping and demoting of txs:
```
[DEBUG] polygon.txpool: dropped account txs: num=519 next_nonce=707 address=0x52adf738AAD93c31f798a30b2C74D658e1E9a562"}
```
```
[DEBUG] polygon.txpool: Demote: threshold reached - dropping account: addr=0x52adf738AAD93c31f798a30b2C74D658e1E9a562"
```
however, the txpool appears to still be stuck when looking at the responses for `txpool_content` and `txpool_inspect`.

As the last successful tx to be written to chain was nonce 707, we would think that tx 707 should be removed from the pending txs, with 708 being included in the next block. This wasn't the case.

We're not sure but suspect this issue could be related to changes introduced in #541.

Lastly, not sure if this is related but but we can observe a different number of dropped txs across several nodes of our core validator set as seen in the additional logs below. Note how some nodes had 521 txs dropped whereas others had 518 txs dropped.

### Your environment

* OS and version `Ubuntu 20`
* version of the Polygon Edge `v0.4.1 (6bb316b1d99c229278076431e6586852ff4e36f9)`
* branch that causes this issue `develop`

### Steps to reproduce

We are trying to find a way to consistently reproduce, but for now here is what happened:

1. Send many eth_call txs from a single account address to a designated RPC cluster (non-validator nodes)
2. Observe that eventually no more txs from the account can be written to chain
3. Observe txpool_inspect and txpool_content reponses that indicate the txpool is stuck on all nodes
4. After restarting all nodes and continuing the service we no longer see the issue

### Expected behaviour

I would expect the txpool tx dropping and demoting logic to properly drop any invalid txs but keep the txpool unblocked.

### Actual behaviour

Somehow the txpool gets blocked and all json-rpc nodes involved required a restart.

### Logs
txpool_inspect response:
```json
{
    "jsonrpc": "2.0",
    "id": null,
    "result": {
        "pending": {
            "0x52adf738AAD93c31f798a30b2C74D658e1E9a562": {
                "707": "0 wei + 5000000 gas x 1232100000 wei",
                "708": "0 wei + 5000000 gas x 1000000000 wei"
            }
        },
        "queued": {
            "0x52adf738AAD93c31f798a30b2C74D658e1E9a562": {
                "708": "0 wei + 5000000 gas x 1000000000 wei",
                "709": "0 wei + 5000000 gas x 1518070410 wei",
                "710": "0 wei + 5000000 gas x 124924000000 wei",
                "711": "0 wei + 5000000 gas x 30039000000 wei",
                "712": "0 wei + 5000000 gas x 1367631000 wei",
                "713": "0 wei + 5000000 gas x 1518070410 wei",
                "714": "0 wei + 5000000 gas x 1367631000 wei",
                "715": "0 wei + 5000000 gas x 1000000000 wei",
                "716": "0 wei + 5000000 gas x 1518070410 wei",
                "717": "0 wei + 5000000 gas x 1367631000 wei",
                "718": "0 wei + 5000000 gas x 1367631000 wei",
                "719": "0 wei + 5000000 gas x 1518070410 wei",
                "720": "0 wei + 5000000 gas x 1367631000 wei",
                "721": "0 wei + 5000000 gas x 2464200000 wei",
                "722": "0 wei + 5000000 gas x 1518070410 wei",
                "723": "0 wei + 5000000 gas x 1110000000 wei",
                "724": "0 wei + 5000000 gas x 1518070410 wei",
                "725": "0 wei + 5000000 gas x 1518070410 wei",
                "726": "0 wei + 5000000 gas x 1518070410 wei",
                "727": "0 wei + 5000000 gas x 1518070410 wei",
                "728": "0 wei + 5000000 gas x 1367631000 wei",
                "729": "0 wei + 5000000 gas x 1518070410 wei",
                "730": "0 wei + 5000000 gas x 1518070410 wei",
                "731": "0 wei + 5000000 gas x 1232100000 wei",
                "732": "0 wei + 5000000 gas x 40290780000 wei",
                "733": "0 wei + 5000000 gas x 101324000000 wei",
                "734": "0 wei + 5000000 gas x 1518070410 wei",
                "735": "0 wei + 5000000 gas x 1232100000 wei",
                "736": "0 wei + 5000000 gas x 1367631000 wei",
                "737": "0 wei + 5000000 gas x 6850355000000 wei",
                "738": "0 wei + 5000000 gas x 1000000000 wei",
                "739": "0 wei + 5000000 gas x 1232100000 wei",
                "740": "0 wei + 5000000 gas x 44722765800 wei",
                "741": "0 wei + 5000000 gas x 49642270038 wei",
                "742": "0 wei + 5000000 gas x 44722765800 wei",
                "743": "0 wei + 5000000 gas x 1232100000 wei",
                "744": "0 wei + 5000000 gas x 40254150000 wei",
                "745": "0 wei + 5000000 gas x 36265000000 wei",
                "746": "0 wei + 5000000 gas x 168206000000 wei",
                "747": "0 wei + 5000000 gas x 1110000000 wei",
                "748": "0 wei + 5000000 gas x 1110000000 wei",
                "749": "0 wei + 5000000 gas x 1232100000 wei",
                "750": "0 wei + 5000000 gas x 1232100000 wei",
                "751": "0 wei + 5000000 gas x 1232100000 wei",
                "752": "0 wei + 5000000 gas x 1232100000 wei",
                "753": "0 wei + 5000000 gas x 1110000000 wei",
                "754": "0 wei + 5000000 gas x 1232100000 wei",
                "755": "0 wei + 5000000 gas x 1232100000 wei",
                "756": "0 wei + 5000000 gas x 1110000000 wei",
                "757": "0 wei + 5000000 gas x 1232100000 wei",
                "758": "0 wei + 5000000 gas x 1232100000 wei",
                "759": "0 wei + 5000000 gas x 1110000000 wei",
                "760": "0 wei + 5000000 gas x 1110000000 wei",
                "761": "0 wei + 5000000 gas x 1110000000 wei",
                "762": "0 wei + 5000000 gas x 1232100000 wei",
                "763": "0 wei + 5000000 gas x 1110000000 wei",
                "764": "0 wei + 5000000 gas x 1110000000 wei",
                "765": "0 wei + 5000000 gas x 1110000000 wei",
                "766": "0 wei + 5000000 gas x 1110000000 wei",
                "767": "0 wei + 5000000 gas x 1000000000 wei",
                "768": "0 wei + 5000000 gas x 1110000000 wei",
                "769": "0 wei + 5000000 gas x 1110000000 wei",
                "770": "0 wei + 5000000 gas x 1000000000 wei",
                "771": "0 wei + 5000000 gas x 1232100000 wei",
                "772": "0 wei + 5000000 gas x 1000000000 wei",
                "773": "0 wei + 5000000 gas x 1232100000 wei",
                "774": "0 wei + 5000000 gas x 1110000000 wei",
                "775": "0 wei + 5000000 gas x 1110000000 wei",
                "776": "0 wei + 5000000 gas x 1232100000 wei",
                "777": "0 wei + 5000000 gas x 1000000000 wei",
                "778": "0 wei + 5000000 gas x 1000000000 wei",
                "779": "0 wei + 5000000 gas x 1000000000 wei",
                "780": "0 wei + 5000000 gas x 1000000000 wei",
                "781": "0 wei + 5000000 gas x 1000000000 wei",
                "782": "0 wei + 5000000 gas x 1000000000 wei",
                "783": "0 wei + 5000000 gas x 1000000000 wei",
                "784": "0 wei + 5000000 gas x 1000000000 wei",
                "785": "0 wei + 5000000 gas x 1000000000 wei",
                "786": "0 wei + 5000000 gas x 1000000000 wei",
                "787": "0 wei + 5000000 gas x 1000000000 wei",
                "788": "0 wei + 5000000 gas x 1000000000 wei",
                "789": "0 wei + 5000000 gas x 1000000000 wei",
                "790": "0 wei + 5000000 gas x 1518070410 wei",
                "791": "0 wei + 5000000 gas x 1518070410 wei",
                "792": "0 wei + 5000000 gas x 1518070410 wei",
                "793": "0 wei + 5000000 gas x 1518070410 wei",
                "794": "0 wei + 5000000 gas x 1518070410 wei",
                "795": "0 wei + 5000000 gas x 1518070410 wei",
                "796": "0 wei + 5000000 gas x 1518070410 wei",
                "797": "0 wei + 5000000 gas x 1518070410 wei",
                "798": "0 wei + 5000000 gas x 1518070410 wei",
                "799": "0 wei + 5000000 gas x 1518070410 wei",
                "800": "0 wei + 5000000 gas x 1518070410 wei",
                "801": "0 wei + 5000000 gas x 1518070410 wei",
                "802": "0 wei + 5000000 gas x 1518070410 wei",
                "803": "0 wei + 5000000 gas x 1518070410 wei",
                "804": "0 wei + 5000000 gas x 1367631000 wei",
                "805": "0 wei + 5000000 gas x 1518070410 wei",
                "806": "0 wei + 5000000 gas x 57460483088 wei",
                "807": "0 wei + 5000000 gas x 1518070410 wei",
                "808": "0 wei + 5000000 gas x 1518070410 wei",
                "809": "0 wei + 5000000 gas x 1518070410 wei",
                "810": "0 wei + 5000000 gas x 1367631000 wei",
                "811": "0 wei + 5000000 gas x 1367631000 wei",
                "812": "0 wei + 5000000 gas x 1367631000 wei",
                "813": "0 wei + 5000000 gas x 1232100000 wei",
                "814": "0 wei + 5000000 gas x 1232100000 wei",
                "815": "0 wei + 5000000 gas x 1232100000 wei",
                "816": "0 wei + 5000000 gas x 1367631000 wei",
                "817": "0 wei + 5000000 gas x 822272000000 wei",
                "818": "0 wei + 5000000 gas x 1232100000 wei",
                "819": "0 wei + 5000000 gas x 1232100000 wei",
                "820": "0 wei + 5000000 gas x 1367631000 wei",
                "821": "0 wei + 5000000 gas x 1232100000 wei",
                "822": "0 wei + 5000000 gas x 1367631000 wei",
                "823": "0 wei + 5000000 gas x 1367631000 wei",
                "824": "0 wei + 5000000 gas x 1367631000 wei",
                "825": "0 wei + 5000000 gas x 1232100000 wei",
                "826": "0 wei + 5000000 gas x 1232100000 wei",
                "827": "0 wei + 5000000 gas x 1232100000 wei",
                "828": "0 wei + 5000000 gas x 912721920000 wei",
                "829": "0 wei + 5000000 gas x 1232100000 wei",
                "830": "0 wei + 5000000 gas x 1367631000 wei",
                "831": "0 wei + 5000000 gas x 1367631000 wei",
                "832": "0 wei + 5000000 gas x 1518070410 wei",
                "833": "0 wei + 5000000 gas x 1518070410 wei",
                "834": "0 wei + 5000000 gas x 1518070410 wei",
                "835": "0 wei + 5000000 gas x 1518070410 wei",
                "836": "0 wei + 5000000 gas x 1367631000 wei",
                "837": "0 wei + 5000000 gas x 1518070410 wei",
                "838": "0 wei + 5000000 gas x 1518070410 wei",
                "839": "0 wei + 5000000 gas x 1518070410 wei",
                "840": "0 wei + 5000000 gas x 1518070410 wei",
                "841": "0 wei + 5000000 gas x 1518070410 wei",
                "842": "0 wei + 5000000 gas x 1518070410 wei",
                "843": "0 wei + 5000000 gas x 1518070410 wei",
                "844": "0 wei + 5000000 gas x 1518070410 wei",
                "845": "0 wei + 5000000 gas x 1518070410 wei",
                "846": "0 wei + 5000000 gas x 1518070410 wei"
            }
        },
        "currentCapacity": 521,
        "maxCapacity": 4096
    }
}
```
`Dropped account txs` and `Demote: threshold reached` logs:
```
{"message":"2022-09-01T19:13:59.020Z [DEBUG] polygon.txpool: dropped account txs: num=518 next_nonce=707 address=0x52adf738AAD93c31f798a30b2C74D658e1E9a562"}
{"message":"2022-09-01T19:13:59.020Z [DEBUG] polygon.txpool: Demote: threshold reached - dropping account: addr=0x52adf738AAD93c31f798a30b2C74D658e1E9a562"}
{"message":"2022-09-01T19:13:57.020Z [DEBUG] polygon.txpool: Demote: threshold reached - dropping account: addr=0x52adf738AAD93c31f798a30b2C74D658e1E9a562"}
{"message":"2022-09-01T19:13:57.020Z [DEBUG] polygon.txpool: dropped account txs: num=520 next_nonce=707 address=0x52adf738AAD93c31f798a30b2C74D658e1E9a562"}
{"message":"2022-09-01T19:13:55.018Z [DEBUG] polygon.txpool: Demote: threshold reached - dropping account: addr=0x52adf738AAD93c31f798a30b2C74D658e1E9a562"}
{"message":"2022-09-01T19:13:55.018Z [DEBUG] polygon.txpool: dropped account txs: num=518 next_nonce=707 address=0x52adf738AAD93c31f798a30b2C74D658e1E9a562"}
{"message":"2022-09-01T19:13:53.017Z [DEBUG] polygon.txpool: Demote: threshold reached - dropping account: addr=0x52adf738AAD93c31f798a30b2C74D658e1E9a562"}
{"message":"2022-09-01T19:13:53.017Z [DEBUG] polygon.txpool: dropped account txs: num=521 next_nonce=707 address=0x52adf738AAD93c31f798a30b2C74D658e1E9a562"}
{"message":"2022-09-01T19:13:51.054Z [DEBUG] polygon.txpool: Demote: threshold reached - dropping account: addr=0x52adf738AAD93c31f798a30b2C74D658e1E9a562"}
{"message":"2022-09-01T19:13:51.054Z [DEBUG] polygon.txpool: dropped account txs: num=518 next_nonce=707 address=0x52adf738AAD93c31f798a30b2C74D658e1E9a562"}
{"message":"2022-09-01T19:13:51.036Z [DEBUG] polygon.txpool: Demote: threshold reached - dropping account: addr=0x52adf738AAD93c31f798a30b2C74D658e1E9a562"}
{"message":"2022-09-01T19:13:51.036Z [DEBUG] polygon.txpool: dropped account txs: num=519 next_nonce=707 address=0x52adf738AAD93c31f798a30b2C74D658e1E9a562"}
```

**Comment by dbrajovic** [comment]
Hey @dankostiuk , thanks for opening the issue.

There was a similar issue (if not the same) with gossip transactions ending up back-to-back with the original one. That was resolved in (#575). I'm guessing you cannot upgrade to `v0.5.0` (due to the memory leak you reported) - would you be okay with us pushing a quick `v0.5.1` release (with the memory fix and all the `v0.5.0` features and fixes) ? 

Or are you comfortable with waiting for a `v0.6.0` release in which the BLS mode would be introduced ?

**Comment by dankostiuk** [comment]
Hey @dbrajovic , exactly - we noticed that #575 could potentially resolve this issue but due to the memory leak and CPU issues present, we were hesitant in upgrading our prod environment to `v0.5.0`. Therefore it would be great if your team pushed out a `v0.5.1` as you suggested. Thanks in advance 🙏

**Comment by mrwillis** [comment]
@dbrajovic @zivkovicmilos roughly, could you estimate when this `v0.5.1` could get released? This is actually a production issue for us currently, and we basically have monitoring to detect this to restart problematic RPC nodes. I rarely use the word ASAP, especially on an open source project, but we'd really appreciate a quick turnaround on this.

**Comment by dbrajovic** [comment]
@dankostiuk @mrwillis 

Apologies for the wait. 
Try out the `release/0.5.1` branch and let us know if any problems manifest.

**Comment by dankostiuk** [comment]
Hey @dbrajovic, @zivkovicmilos ,

Unfortunately we are still seeing similar errors on our prod environment after pulling in `release/0.5.1`. At the moment we are trying to determine a consistent way of reproducing the issue but for now we think it either involves:

1. how the queued txpool txs are gossiped which cause discrepancies in txpool_inspect responses across our RPC cluster nodes, e.g. not all queued nonces match for an account when querying different nodes
2. something strange occurring when txs with nonces higher than the current nonce are queued up (5516), followed by a series of txs being mined from the current nonce up until the same queued nonce (5500-5516) which results in the tx with nonce 5516 not properly being promoted (or the nonce 5516 in pending not properly getting cleared):
```
{
    "jsonrpc": "2.0",
    "id": null,
    "result": {
        "pending": {
            "0x52adf738AAD93c31f798a30b2C74D658e1E9a562": {
                "5516": "0 wei + 5000000 gas x 1367631000 wei"
            }
        },
        "queued": {
            "0x52adf738AAD93c31f798a30b2C74D658e1E9a562": {
                "5516": "0 wei + 5000000 gas x 1232100000 wei",
                "5517": "0 wei + 5000000 gas x 1000000000 wei",
                "5518": "0 wei + 5000000 gas x 1000000000 wei",
                "5519": "0 wei + 5000000 gas x 1000000000 wei",
                "5520": "0 wei + 5000000 gas x 1000000000 wei"
            }
        },
        "currentCapacity": 7,
        "maxCapacity": 4096
    }
}
```

During this time we've seen errors of:
```
[ERROR] polygon.txpool: failed to add tx: err="already known"
```
and
```
[ERROR] polygon.txpool: failed to add tx: err="nonce too low"
```

**Comment by mrwillis** [comment]
Would say this is a pretty critical bug as it can effectively brick accounts on the chain until txpools are refreshed across nodes.

**Comment by dankostiuk** [comment]
Was able to repro what we believe to be a similar scenario to the one we initially observed using the steps below:

1. assuming account balance is 5 ETH - submit a couple simple transfer txs we can currently afford with current nonce + 10 (for example each tx costing total of 3 ETH each), these txs will then appear within `queued` in txpool_inspect response (for example with nonces 450-451)
2. submit 10 txs we can currently afford using the current nonce (for example nonces 440-449 with each tx costing only 0.1 ETH each)
3. once 449 has been queued, promoted and mined, the queued txs with nonces 450 and 451 will be promoted to pending however at this point, the account balance cannot cover the tx costs:
```
{
    "jsonrpc": "2.0",
    "id": null,
    "result": {
        "pending": {
            "0xFd9E636956415cC447Fb621223d2fc779819E533": {
                "450": "3000000000000000000 wei + 21000 gas x 10000000000 wei",
                "451": "3000000000000000000 wei + 21000 gas x 10000000000 wei"
            }
        },
        "queued": {},
        "currentCapacity": 2,
        "maxCapacity": 4096
    }
}
```
4. these stuck pending txs do not go away on their own even once the account balance is topped up
---

Not sure but I would expect the queued txs to consider the account balance at the latest block during the promotion process and therefore not get promoted.

**Comment by lazartravica** [comment]
@mrwillis @dankostiuk thanks for letting us know. We are tackling both of these immediately, with the original reported one first. 

Let's coordinate about the second issue per @dankostiuk's last comment in the separate issue thread: https://github.com/0xPolygon/polygon-edge/issues/731

**Comment by dbrajovic** [comment]
@mrwillis @dankostiuk 

After a closer look at the [result](https://github.com/0xPolygon/polygon-edge/issues/711#issuecomment-1243985582) you posted, we believe the core of the bug is identified and a fix is under way.

It is a combination of a faulty [promote](https://github.com/0xPolygon/polygon-edge/blob/68df30d52da8a416b9c7c5860372f951cc72e325/txpool/account.go#L246) logic and an unlucky timing of `ResetWithHeaders`. 

We should never see double nonces in promoted queue, but in this particular example:
```
    "result": {
        "pending": {
            "0x52adf738AAD93c31f798a30b2C74D658e1E9a562": {
                "5516": "0 wei + 5000000 gas x 1367631000 wei"   // note the wei
            }
        },
        "queued": {
            "0x52adf738AAD93c31f798a30b2C74D658e1E9a562": {
                "5516": "0 wei + 5000000 gas x 1232100000 wei",  // here as well
                // ...
            }
        },

```

the one in `queued` is not from gossip, but an entirely different transaction that shares the same nonce. Everything else that was reported comes out as a consequence of this event. 

On the other hand [this](https://github.com/0xPolygon/polygon-edge/issues/711#issuecomment-1244680548) looks like another issue which demands further inquiry. We'll pick up conversation on that [here](https://github.com/0xPolygon/polygon-edge/issues/731)

**Comment by Kourin1996** [comment]
The above PR is going to be included at 0.5.1 release
